### PR TITLE
Add mutliple values for data/args/rets

### PR DESCRIPTION
**Pull request by nezuo**
TODO:
- [x] TypeScript output
- [x] Tooling output
- [x] Docs

### CODE REVIEW

**Comment by nezuo** [comment]
PR is ready for review.

File: zap/src/irgen/des.rs [hunk 1]
```diff
@@ -14,8 +14,20 @@ impl Gen for Des {
...
+			let var = if i > 0 {
+				Var::Name(match &var {
+					Var::Name(name) => format!("{name}{}", i + 1),
+					_ => unreachable!(),
+				})
+			} else {
+				var.clone()
+			};
+
+			self.push_ty(ty, var);
```
Comment (sasial-dev) [inline zap/src/irgen/des.rs:28]: This is... awkward.
Let's instead add an `nsuffix` method to the `Var` impl (see `var.nindex` as a good example).
And then we can do something like
```suggestion
			self.push_ty(ty, var.nsuffix(i + 1));
```

File: zap/src/irgen/ser.rs [hunk 2]
```diff
@@ -14,8 +14,20 @@ impl Gen for Ser {
...
+			let var = if i > 0 {
+				Var::Name(match &var {
+					Var::Name(name) => format!("{name}{}", i + 1),
+					_ => unreachable!(),
+				})
+			} else {
+				var.clone()
+			};
+
+			self.push_ty(ty, var);
```
Comment (sasial-dev) [inline zap/src/irgen/ser.rs:28]: Same comment as above

File: zap/src/output/luau/client.rs [hunk 3]
```diff
@@ -174,6 +174,23 @@ impl<'src> ClientOutput<'src> {
...
+	fn get_values(&self, data: &Option<Vec<Ty>>) -> String {
+		if let Some(types) = data {
+			(1..=types.len())
+				.map(|i| {
+					if i == 1 {
+						"value".to_string()
+					} else {
+						format!("value{}", i)
+					}
+				})
```
Comment (sasial-dev) [inline zap/src/output/luau/client.rs:186]: If you apply my comment above:
```suggestion
				.map(|i| format!("value{}", i))
```

File: zap/src/output/luau/client.rs [hunk 4]
```diff
@@ -174,6 +174,23 @@ impl<'src> ClientOutput<'src> {
...
+				.collect::<Vec<_>>()
+				.join(", ")
+		} else {
+			"value".to_string()
```
Comment (sasial-dev) [inline zap/src/output/luau/client.rs:190]:
```suggestion
			"value1".to_string()
```

File: zap/src/output/luau/client.rs [hunk 5]
```diff
@@ -515,16 +548,32 @@ impl<'src> ClientOutput<'src> {
 		));
 	}
 
+	fn push_value_parameters(&mut self, types: &[Ty]) {
+		for (i, ty) in types.iter().enumerate() {
+			let value = format!(
+				"{}{}",
+				self.config.casing.with("Value", "value", "value"),
+				if i == 0 { "".to_string() } else { (i + 1).to_string() }
```
Comment (sasial-dev) [inline zap/src/output/luau/client.rs:556]:
```suggestion
				(i + 1).to_string()
```

File: zap/src/output/typescript/client.rs [hunk 6]
```diff
@@ -64,6 +65,22 @@ impl<'src> ClientOutput<'src> {
...
+		let value = self.config.casing.with("Value", "value", "value");
+
+		for (i, ty) in types.iter().enumerate() {
+			if i > 0 {
+				self.push(", ");
+			}
+
+			self.push(&format!(
+				"{value}{}",
+				if i > 0 { (i + 1).to_string() } else { "".to_string() },
```
Comment (sasial-dev) [inline zap/src/output/typescript/client.rs]:
```suggestion
				(i + 1).to_string(),
```

File: zap/src/output/typescript/client.rs [hunk 7]
```diff
@@ -64,6 +65,22 @@ impl<'src> ClientOutput<'src> {
 		}
 	}
 
+	fn push_parameters(&mut self, types: &[Ty]) {
```
Comment (sasial-dev) [inline zap/src/output/typescript/client.rs]: This function should maybe be in `mod.rs`?

File: zap/src/output/luau/server.rs [hunk 8]
```diff
@@ -259,12 +281,36 @@ impl<'a> ServerOutput<'a> {
...
 
+		let rets = if let Some(types) = &fndecl.args {
+			(1..=types.len())
+				.map(|i| {
+					if i > 1 {
+						format!("rets{}", i)
+					} else {
+						"rets".to_string()
+					}
+				})
```
Comment (sasial-dev) [inline zap/src/output/luau/server.rs:292]:
```suggestion
				.map(|i| format!("rets{}", i))
```

File: zap/src/output/luau/server.rs [hunk 9]
```diff
@@ -447,8 +495,25 @@ impl<'a> ServerOutput<'a> {
 		));
 	}
 
+	fn push_value_parameters(&mut self, types: &[Ty]) {
+		for (i, ty) in types.iter().enumerate() {
+			let value = format!(
+				"{}{}",
+				self.config.casing.with("Value", "value", "value"),
+				if i == 0 { "".to_string() } else { (i + 1).to_string() }
```
Comment (sasial-dev) [inline zap/src/output/luau/server.rs:503]:
```suggestion
				(i + 1).to_string()
```

**Review by sasial-dev — CHANGES_REQUESTED** [review]
Thank you for this!
It pretty much looks good to me, but it would be nice to address these minor things :)

**Review by sasial-dev — APPROVED** [review]
LGTM - thank you!

**Pull request merged**